File: backend/app/extraction/schemas.py
"""Pydantic schemas — source of truth. Hand-mirrored in frontend/src/lib/schemas.ts."""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class SeniorityLevel(str, Enum):
    junior = "junior"
    mid = "mid"
    senior = "senior"
    lead = "lead"
    unknown = "unknown"


class ExtractedProfile(BaseModel):
    skills: list[str] = Field(default_factory=list)
    years_experience: Optional[float] = None
    seniority_level: Optional[SeniorityLevel] = None
    roles: list[str] = Field(default_factory=list)
    education: list[str] = Field(default_factory=list)
    summary: str = ""
    # Completeness signals — set by extractor, not the LLM
    extraction_warnings: list[str] = Field(default_factory=list)


class ExtractedRequirements(BaseModel):
    required_skills: list[str] = Field(default_factory=list)
    nice_to_have_skills: list[str] = Field(default_factory=list)
    min_years_experience: Optional[float] = None
    seniority_level: Optional[SeniorityLevel] = None
    responsibilities: list[str] = Field(default_factory=list)
    # Exact phrases for ATS keyword check (Day 3); populated during JD extraction
    ats_phrases: list[str] = Field(default_factory=list)
    extraction_warnings: list[str] = Field(default_factory=list)


class MatchConfidence(str, Enum):
    matched = "matched"
    review = "review"
    missing = "missing"


class SkillMatch(BaseModel):
    resume_skill: Optional[str] = None
    jd_skill: str
    confidence: MatchConfidence
    similarity: Optional[float] = None
    reason: Optional[str] = None


class CategoryScores(BaseModel):
    technical: float = 0.0
    experience: float = 0.0
    domain: float = 0.0
    soft: float = 0.0


class AtsFlag(BaseModel):
    phrase: str
    found_in_resume: bool


class RewriteSuggestion(BaseModel):
    original: str
    suggested: str
    targets_skill: str
    rationale: str


class AnalyzeResponse(BaseModel):
    overall_score: float = Field(ge=0, le=100)
    category_scores: CategoryScores = Field(default_factory=CategoryScores)
    matched_skills: list[SkillMatch] = Field(default_factory=list)
    missing_required: list[SkillMatch] = Field(default_factory=list)
    missing_nice_to_have: list[SkillMatch] = Field(default_factory=list)
    overqualified: list[str] = Field(default_factory=list)
    review_band: list[SkillMatch] = Field(default_factory=list)
    summary: str = ""
    profile: Optional[ExtractedProfile] = None
    requirements: Optional[ExtractedRequirements] = None
    ats_flags: list[AtsFlag] = Field(default_factory=list)
    rewrite_suggestions: list[RewriteSuggestion] = Field(default_factory=list)
    # Day 1: when match/score is stubbed
    matching_stubbed: bool = False

File: frontend/src/lib/schemas.ts
/**
 * Zod schemas hand-mirrored from backend/app/extraction/schemas.py (Pydantic SoT).
 */
import { z } from "zod";

export const SeniorityLevelSchema = z.enum([
  "junior",
  "mid",
  "senior",
  "lead",
  "unknown",
]);

export const ExtractedProfileSchema = z.object({
  skills: z.array(z.string()),
  years_experience: z.number().nullable().optional(),
  seniority_level: SeniorityLevelSchema.nullable().optional(),
  roles: z.array(z.string()),
  education: z.array(z.string()),
  summary: z.string(),
  extraction_warnings: z.array(z.string()).default([]),
});

export const ExtractedRequirementsSchema = z.object({
  required_skills: z.array(z.string()),
  nice_to_have_skills: z.array(z.string()),
  min_years_experience: z.number().nullable().optional(),
  seniority_level: SeniorityLevelSchema.nullable().optional(),
  responsibilities: z.array(z.string()),
  ats_phrases: z.array(z.string()).default([]),
  extraction_warnings: z.array(z.string()).default([]),
});

export const MatchConfidenceSchema = z.enum(["matched", "review", "missing"]);

export const SkillMatchSchema = z.object({
  resume_skill: z.string().nullable().optional(),
  jd_skill: z.string(),
  confidence: MatchConfidenceSchema,
  similarity: z.number().nullable().optional(),
  reason: z.string().nullable().optional(),
});

export const CategoryScoresSchema = z.object({
  technical: z.number(),
  experience: z.number(),
  domain: z.number(),
  soft: z.number(),
});

export const AtsFlagSchema = z.object({
  phrase: z.string(),
  found_in_resume: z.boolean(),
});

export const RewriteSuggestionSchema = z.object({
  original: z.string(),
  suggested: z.string(),
  targets_skill: z.string(),
  rationale: z.string(),
});

export const AnalyzeResponseSchema = z.object({
  overall_score: z.number().min(0).max(100),
  category_scores: CategoryScoresSchema,
  matched_skills: z.array(SkillMatchSchema),
  missing_required: z.array(SkillMatchSchema),
  missing_nice_to_have: z.array(SkillMatchSchema),
  overqualified: z.array(z.string()),
  review_band: z.array(SkillMatchSchema),
  summary: z.string(),
  profile: ExtractedProfileSchema.nullable().optional(),
  requirements: ExtractedRequirementsSchema.nullable().optional(),
  ats_flags: z.array(AtsFlagSchema).default([]),
  rewrite_suggestions: z.array(RewriteSuggestionSchema).default([]),
  matching_stubbed: z.boolean().default(false),
});

export const JdComparisonItemSchema = z.object({
  label: z.string(),
  overall_score: z.number().min(0).max(100),
  summary: z.string(),
  top_gaps: z.array(z.string()),
  result: AnalyzeResponseSchema,
});

export const CompareResponseSchema = z.object({
  ranked: z.array(JdComparisonItemSchema),
  profile: ExtractedProfileSchema.nullable().optional(),
});

export type AnalyzeResponse = z.infer<typeof AnalyzeResponseSchema>;
export type CompareResponse = z.infer<typeof CompareResponseSchema>;
export type JdComparisonItem = z.infer<typeof JdComparisonItemSchema>;
export type SkillMatch = z.infer<typeof SkillMatchSchema>;
export type CategoryScores = z.infer<typeof CategoryScoresSchema>;
export type AtsFlag = z.infer<typeof AtsFlagSchema>;
export type RewriteSuggestion = z.infer<typeof RewriteSuggestionSchema>;
